### DISCUSSION TITLE
Build "dependency" tree UI

### DISCUSSION THREAD

**Discussion by jmikrut** [payloadcms/payload#2938] [Roadmap]
We need to build a way to manage relationship / upload "dependency trees".

This would work in two ways:

#### 1 - relationships TO a document

We should surface all relationships TO a given document when you navigate to the document. For example, if you are viewing a media file, you should see all documents that relate to that given media document when you are viewing the media document itself in the admin UI. This will open up the possibility to, for example, prompt the user upon deletion of the document to say something like "You are about to delete this media document, but these X number of documents relate to the media document." 

#### 2 - Relationships ON a document

We should also surface all relationships that a parent document has relations to. For example, if you are editing a post, and it has relationships to X number of media files, and X number of categories, tags, etc - - we should be able to view a list of all dependent relations easily. This will be helpful in cases where, say, you want to "bulk publish" changes to a parent post, but also want to bulk publish changes to all related documents at the same time.

Both of these new features in conjunction will allow us to build dynamic caching / purging down the line as well. For example, let's take an example REST response at `/api/posts/:my-id?depth=2`. If someone wanted to cache this URL, we would need to be able to purge this URL from cache **_not only when the post itself changes,_** but also when any of its relations up to 2 levels deep change.

**Comment by DanRibbens** [comment]
Related to https://github.com/payloadcms/payload/discussions/3008

**Comment by jacobsfletch** [comment]
Design in #3217

**Comment by vanska** [comment]
I've already built this functionality for plugins that:

- incrementally revalidate cache tags when references change no matter how deeply nested
- guard documents against deletion while still being referenced

Some issues/friction I've faced along the way:

- https://github.com/payloadcms/payload/discussions/15280
- https://github.com/payloadcms/payload/discussions/2100#discussioncomment-15550053

**Comment by GermanJablo** [comment]
Done a long time ago
